src/settingsSync.ts

```diff
@@ -16,7 +35,75 @@ interface ConfigurationSyncStore {
 	authenticationProviders: Record<string, { scopes: string[] }>;
 }
 
-export default class SettingsSync extends Disposable {
+export const enum SyncResource {
+	Settings = 'settings',
+	Keybindings = 'keybindings',
+	Snippets = 'snippets',
+	Tasks = 'tasks',
+	Extensions = 'extensions',
+	GlobalState = 'globalState',
+}
+
+export interface IExtensionIdentifier {
+	id: string;
+	uuid?: string;
+}
+
+export interface ISyncExtension {
+	identifier: IExtensionIdentifier;
+	preRelease?: boolean;
+	version?: string;
+	disabled?: boolean;
+	installed?: boolean;
+	state?: Record<string, any>;
+}
+
+export interface ISyncData {
+	version: number;
+	machineId?: string;
+	content: string;
+}
+
+export function isSyncData(thing: any): thing is ISyncData {
+	if (thing
+		&& (thing.version !== undefined && typeof thing.version === 'number')
+		&& (thing.content !== undefined && typeof thing.content === 'string')) {
+
+		// backward compatibility
+		if (Object.keys(thing).length === 2) {
+			return true;
+		}
+
+		if (Object.keys(thing).length === 3
+			&& (thing.machineId !== undefined && typeof thing.machineId === 'string')) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+export function parseSyncData(content: string): ISyncData | undefined {
+	try {
+		const syncData: ISyncData = JSON.parse(content);
+		if (isSyncData(syncData)) {
+			return syncData;
+		}
+	} catch {
```

Comment:
should we log like source did?

Reply:
I'll add a log but I'd expect this to always work, if for some reason it failed then it would be mean there's a bigger issue as that would mean settings sync is broken in vscode web too

Reply:
Ah this is logged in `remoteConnectort.ts` line 951, as it should never be undefined